File: java/data-structures-algorithms-samples/src/main/java/me/jkong/data/queue/Queue.java
package me.jkong.data.queue;

/**
 * @author JKong
 * @version v1.0
 * @description queue
 * @date 2019/12/21 9:27 下午.
 */
public interface Queue<E> {
    
    /**
     * 队列中添加元素
     *
     * @param e 元素
     */
    void enqueue(E e);
    
    /**
     * 从队列中取出元素
     *
     * @return 元素
     */
    E dequeue();
    
    /**
     * 获取队列前面的一个元素
     *
     * @return 元素
     */
    E getFront();
    
    /**
     * 获取队列的大小
     *
     * @return 大小
     */
    int getSize();
    
    /**
     * 队列是否为空
     *
     * @return true：空，false：不为空
     */
    boolean isEmpty();
    
}

File: go/src/me.jkong/go/hello-go-sample/04_oop/type/queue/Queue.go
// @Description: 数据结构 queue
// @Author: JKong
// @Update: 2020/10/2 5:22 下午
package queue

import "fmt"

// 定义数据类型 queue
type Queue []int

// 设置方法

// 添加元素
func (q *Queue) Push(v int) {
	*q = append(*q, v)
}

// 移除元素
func (q *Queue) Pop() int {
	head := (*q)[0]
	*q = (*q)[1:]
	return head
}

// 是否为空
func (q *Queue) IsEmpty() bool {
	return len(*q) == 0
}

// 打印
func (q *Queue) ToString() {
	fmt.Print("[")
	for _, v := range *q {
		fmt.Print(v, ", ")
	}
	fmt.Print("]\n")
}
